### ISSUE TITLE
Copy of Upgrade document-download-api to Python 3.12

### ISSUE THREAD

**Issue by sastels** [cds-snc/notification-planning-core#464]
# Description

As a software developer working on GCNotify stack,
I want the Python stack to be upgraded at latest supported version,
So that I can leverage the latest features of the language and better/safer code.

As a system ops working on GCNotify stack,
I want the Python stack to be upgraded at latest supported version,
So that I can leverage the latest performance improvement
And process notifications faster.

## WHY are we building?

To get latest performance improvements, features and security updates of the Python language/runtime.

## WHAT are we building?

Upgrade the stack to the latest possible version of Python. A few considerations:

* It is possible that we cannot upgrade to the latest of the latest (3.12 at the time of this writing for the stable version), as this might be blocked by a dependency library that did not release a version supporting the latest. For example, we might have to upgrade from version 3.10 to version 3.11, and wait a bit longer to migrate to version 3.12.
* The lambdas might have additional considerations that might limit the possible maximum Python version we can use, such as supported runtime and also monitoring libraries that need to support that environment (such as New Relic). Consider doing this work in #354 .

## VALUE created by our solution

Better performance, security and reliability.

# Acceptance Criteria

- [ ] Document download api running on Python 3.12.
- [ ] test locally
- [ ] test on dev
- [ ] document in tricks and tips how we can deploy branches to dev

# QA Steps

- [ ] Run smoke tests
- [ ] bug bash on staging

**Comment by jimleroyer** [comment]
Steve found new tests in admin that were failing, triggered by the Python 3.12 upgrade. He fixed some but there might be some remaining.

**Comment by jimleroyer** [comment]
Steve fixed a few tests involving a specific Redis command that flowed down to the API command.

**Comment by sastels** [comment]
:/ It really wasn't a redis command problem, instead the problem that my poetry.lock file was not installing the latest version of the 3.12 utils branch. All good now though.

**Comment by sastels** [comment]
ok, have a PR to merge the app team branch into main's 3.12
https://github.com/cds-snc/notification-utils/pull/341/

Using this new PR branch when testing the new api 3.12 PR.

**Comment by sastels** [comment]
moving back to blocked until we finish testing the 3.12 api / celery.

**Comment by jimleroyer** [comment]
This is in production. We will do a bug bash today to make sure everything works as expected.

**Comment by jimleroyer** [comment]
Bug bash was ran yesterday and reported no issues. Closing the issue.
